File: scgraph/helpers/kd_tree.py
from math import pi, cos, sin


def kdtree(points, depth=0, axis_count=2):
    """
    Function:

    - Build a KDTree from a list of points.

    Required Arguments:

    - `points`
        - Type: list of tuples
        - What: A list of points to build the KDTree from

    Optional Arguments:

    - `depth`
        - Type: int
        - What: The current depth in the tree (used for axis selection)
    - `axis_count`
        - Type: int
        - What: The number of dimensions in the points (default is 2 for 2D points)

    Returns:

    - The constructed KDTree as a tuple in the format (point, axis, left, right).
    - Where left and right are subtrees.
    """
    if not points:
        return 0
    axis = depth % axis_count
    points.sort(key=lambda p: p[axis])
    median = len(points) // 2
    return (
        points[median],
        axis,
        kdtree(points[:median], depth + 1),
        kdtree(points[median + 1 :], depth + 1),
    )


def squared_distance(p1, p2, axis_count=2):
    """
    Function:

    - Calculate the squared distance between two points.

    Required Arguments:

    - `p1`
        - Type: tuple
        - What: The first point
    - `p2`
        - Type: tuple
        - What: The second point

    Optional Arguments:

    - `axis_count`
        - Type: int
        - What: The number of dimensions in the points
        - Default: 2 (for 2D points)

    Returns:

    - The squared distance between the two points.
    """
    return sum([(p1[i] - p2[i]) ** 2 for i in range(axis_count)])


def closest_point(
    node, point, depth=0, best=None, axis_count=2, best_dist=float("inf")
):
    """
    Function:

    - Find the closest point in the KDTree to a given point.

    Required Arguments:

    - `node`
        - Type: tuple
        - What: The node of the KDTree
    - `point`
        - Type: tuple
        - What: The point to find the closest point to

    Optional Arguments:

    - `depth`
        - Type: int
        - What: The current depth in the tree (used for axis selection)
    - `best`
        - Type: tuple or None
        - What: The best point found so far (default is None)
    - `axis_count`
        - Type: int
        - What: The number of dimensions in the points (default is 2 for 2D points)
    - `best_dist`
        - Type: float
        - What: The best distance found so far (default is infinity)

    Returns:

    - The closest point found in the KDTree to the given point.
    """
    if node == 0:
        return best, best_dist
    # Get the median node and its distance
    median_node = node[0]
    median_node_dist = squared_distance(
        point, median_node, axis_count=axis_count
    )
    # Update the best point and distance if necessary
    if best is None or median_node_dist < best_dist:
        best = median_node
        best_dist = median_node_dist
    # Calculate the difference for node selection given the current axis
    axis = node[1]
    diff = point[axis] - median_node[axis]
    # Choose side to search
    close, away = (node[2], node[3]) if diff < 0 else (node[3], node[2])
    # Search the close side first
    best, best_dist = closest_point(
        close,
        point,
        depth + 1,
        best,
        axis_count=axis_count,
        best_dist=best_dist,
    )
    # Check the other side if needed
    if diff**2 < best_dist:
        best, best_dist = closest_point(
            away,
            point,
            depth + 1,
            best,
            axis_count=axis_count,
            best_dist=best_dist,
        )
    return best, best_dist


def squared_distance_3d(p1, p2):
    """
    Function:

    - Calculate the squared distance between two 3D points.

    Required Arguments:

    - `p1`
        - Type: tuple
        - What: The first point in 3D space
    - `p2`
        - Type: tuple
        - What: The second point in 3D space

    Returns:

    - The squared distance between the two 3D points.
    """
    return sum(
        [(p1[0] - p2[0]) ** 2, (p1[1] - p2[1]) ** 2, (p1[2] - p2[2]) ** 2]
    )


def closest_point_3d(node, point, depth=0, best=None, best_dist=float("inf")):
    """
    Function:

    - Find the closest point in the KDTree to a given point.

    Required Arguments:

    - `node`
        - Type: tuple
        - What: The node of the KDTree
    - `point`
        - Type: tuple
        - What: The point to find the closest point to

    Optional Arguments:

    - `depth`
        - Type: int
        - What: The current depth in the tree (used for axis selection)
    - `best`
        - Type: tuple or None
        - What: The best point found so far (default is None)
    - `best_dist`
        - Type: float
        - What: The best distance found so far (default is infinity)

    Returns:

    - The closest point found in the KDTree to the given point.
    """
    if node == 0:
        return best, best_dist
    # Get the median node and its distance
    median_node = node[0]
    median_node_dist = squared_distance_3d(point, median_node)
    # Update the best point and distance if necessary
    if best is None or median_node_dist < best_dist:
        best = median_node
        best_dist = median_node_dist
    # Calculate the difference for node selection given the current axis
    axis = node[1]
    diff = point[axis] - median_node[axis]
    # Choose side to search
    close, away = (node[2], node[3]) if diff < 0 else (node[3], node[2])
    # Search the close side first
    best, best_dist = closest_point_3d(close, point, depth + 1, best, best_dist)
    # Check the other side if needed
    if diff**2 < best_dist:
        best, best_dist = closest_point_3d(
            away, point, depth + 1, best, best_dist
        )
    return best, best_dist


class KDTree:
    def __init__(self, points):
        """
        Function:

        - Build a KDTree from a list of points.

        Required Arguments:

        - `points`
            - Type: list of tuples
            - What: A list of points to build the KDTree from

        Returns:

        - A KDTree object that can be used to find the closest point to a given point.
        """
        self.tree = kdtree(points, axis_count=len(points[0]))

    def closest_point(self, point):
        """
        Function:

        - Find the closest point in the KDTree to a given point.

        Required Arguments:

        - `point`
            - Type: tuple
            - What: The point to find the closest point to

        Returns:

        - The closest point found in the KDTree to the given point.
        """
        return closest_point(self.tree, point)[
            0
        ]  # Return only the point, not the distance


class GeoKDTree:
    def __init__(self, points: list[tuple]):
        """
        Function:

        - Build a GeoKDTree from a list of latitude and longitude points or an existing KDTree.

        Required Arguments:

        - `points`
            - Type: list of tuples
            - What: A list of latitude and longitude points to build the GeoKDTree from
            - The points should be in the format [(lat1, lon1), (lat2, lon2), ...].

        Returns:

        - A GeoKDTree object that can be used to find the closest point to a given latitude and longitude.
        - The points should be in the format [(lat1, lon1), (lat2, lon2), ...].
        """
        self.tree = kdtree(
            [
                GeoKDTree.lat_lon_idx_to_xyz_idx(point[0], point[1], idx)
                for idx, point in enumerate(points)
            ],
            axis_count=3,
        )

    def closest_idx(self, point: tuple):
        """
        Function:

        - Find the index of the closest point in the GeoKDTree to a given latitude and longitude point.

        Required Arguments:

        - `point`
            - Type: tuple
            - What: The latitude and longitude point to find the closest point to

        Returns:

        - The index of the closest point found in the GeoKDTree to the given latitude and longitude point.
        """
        return closest_point_3d(
            self.tree,
            GeoKDTree.lat_lon_idx_to_xyz_idx(point[0], point[1]),
        )[0][
            3
        ]  # Return the point [0] and the index of the point [3]

    @staticmethod
    def lat_lon_idx_to_xyz_idx(
        lat: int | float, lon: int | float, idx: int = 0
    ):
        """
        Function:

        - Convert latitude and longitude to Cartesian coordinates (x, y, z) and include an index.

        Required Arguments:

        - `lat`
            - Type: int or float
            - What: The latitude in degrees
        - `lon`
            - Type: int or float
            - What: The longitude in degrees

        Optional Arguments:

        - `idx`
            - Type: int
            - What: An index to include with the coordinates (default is 0)
        """
        lat_rad = lat * pi / 180
        lon_rad = lon * pi / 180
        cos_lat = cos(lat_rad)
        x = cos_lat * cos(lon_rad)
        y = cos_lat * sin(lon_rad)
        z = sin(lat_rad)
        return (x, y, z, idx)

File: cpp/src/kdtree.cpp
#include "kdtree.hpp"
#include <cmath>
#include <algorithm>
#include <stdexcept>

namespace kdtree_helpers {
    double squared_distance(const std::vector<double>& p1, const std::vector<double>& p2, int axis_count) {
        double sum = 0.0;
        for (int i = 0; i < axis_count; ++i) {
            double diff = p1[i] - p2[i];
            sum += diff * diff;
        }
        return sum;
    }
    
    double squared_distance_3d(const std::vector<double>& p1, const std::vector<double>& p2) {
        double dx = p1[0] - p2[0];
        double dy = p1[1] - p2[1];
        double dz = p1[2] - p2[2];
        return dx * dx + dy * dy + dz * dz;
    }
    
    std::vector<double> lat_lon_idx_to_xyz_idx(double lat, double lon, int idx) {
        constexpr double PI = 3.14159265358979323846;
        double lat_rad = lat * PI / 180.0;
        double lon_rad = lon * PI / 180.0;
        double cos_lat = std::cos(lat_rad);
        double x = cos_lat * std::cos(lon_rad);
        double y = cos_lat * std::sin(lon_rad);
        double z = std::sin(lat_rad);
        return {x, y, z, static_cast<double>(idx)};
    }
}

// ============================================================================
// KDTree Implementation
// ============================================================================

KDTree::KDTree(const std::vector<std::vector<double>>& points) {
    if (points.empty()) {
        throw std::invalid_argument("Cannot build KDTree from empty point list");
    }
    
    dimensions = points[0].size();
    
    // Make a mutable copy of points for sorting
    std::vector<std::vector<double>> points_copy = points;
    
    root = build_tree(points_copy, 0, dimensions);
}

std::shared_ptr<KDNode> KDTree::build_tree(
    std::vector<std::vector<double>>& points, 
    int depth, 
    int axis_count
) {
    if (points.empty()) {
        return nullptr;
    }
    
    int axis = depth % axis_count;
    
    // Sort points by the current axis
    std::sort(points.begin(), points.end(),
        [axis](const std::vector<double>& a, const std::vector<double>& b) {
            return a[axis] < b[axis];
        }
    );
    
    size_t median = points.size() / 2;
    
    // Create node with median point
    auto node = std::make_shared<KDNode>(points[median], axis);
    
    // Build left and right subtrees
    if (median > 0) {
        std::vector<std::vector<double>> left_points(points.begin(), points.begin() + median);
        node->left = build_tree(left_points, depth + 1, axis_count);
    }
    
    if (median + 1 < points.size()) {
        std::vector<std::vector<double>> right_points(points.begin() + median + 1, points.end());
        node->right = build_tree(right_points, depth + 1, axis_count);
    }
    
    return node;
}

void KDTree::find_closest(
    const std::shared_ptr<KDNode>& node,
    const std::vector<double>& point,
    int depth,
    std::vector<double>& best,
    double& best_dist,
    int axis_count
) const {
    if (!node) {
        return;
    }
    
    // Calculate distance to current node
    double node_dist = kdtree_helpers::squared_distance(point, node->point, axis_count);
    
    // Update best if necessary
    if (best.empty() || node_dist < best_dist) {
        best = node->point;
        best_dist = node_dist;
    }
    
    // Determine which side to search first
    int axis = node->axis;
    double diff = point[axis] - node->point[axis];
    
    auto close = (diff < 0) ? node->left : node->right;
    auto away = (diff < 0) ? node->right : node->left;
    
    // Search the close side first
    find_closest(close, point, depth + 1, best, best_dist, axis_count);
    
    // Check if we need to search the other side
    if (diff * diff < best_dist) {
        find_closest(away, point, depth + 1, best, best_dist, axis_count);
    }
}

std::vector<double> KDTree::closest_point(const std::vector<double>& point) const {
    if (!root) {
        throw std::runtime_error("KDTree is empty");
    }
    
    std::vector<double> best;
    double best_dist = std::numeric_limits<double>::infinity();
    
    find_closest(root, point, 0, best, best_dist, dimensions);
    
    return best;
}

ClosestPointResult KDTree::closest_point_with_distance(const std::vector<double>& point) const {
    if (!root) {
        throw std::runtime_error("KDTree is empty");
    }
    
    std::vector<double> best;
    double best_dist = std::numeric_limits<double>::infinity();
    
    find_closest(root, point, 0, best, best_dist, dimensions);
    
    return ClosestPointResult{best, best_dist};
}

// ============================================================================
// GeoKDTree Implementation
// ============================================================================

GeoKDTree::GeoKDTree(const std::vector<std::pair<double, double>>& points) {
    if (points.empty()) {
        throw std::invalid_argument("Cannot build GeoKDTree from empty point list");
    }
    
    // Convert lat/lon points to xyz coordinates
    std::vector<std::vector<double>> xyz_points;
    xyz_points.reserve(points.size());
    
    for (size_t i = 0; i < points.size(); ++i) {
        xyz_points.push_back(
            kdtree_helpers::lat_lon_idx_to_xyz_idx(
                points[i].first,
                points[i].second,
                i
            )
        );
    }
    
    root = build_tree(xyz_points, 0);
}

std::shared_ptr<GeoKDNode> GeoKDTree::build_tree(
    std::vector<std::vector<double>>& points,
    int depth
) {
    if (points.empty()) {
        return nullptr;
    }
    
    int axis = depth % 3;  // 3D for x, y, z
    
    // Sort points by the current axis
    std::sort(points.begin(), points.end(),
        [axis](const std::vector<double>& a, const std::vector<double>& b) {
            return a[axis] < b[axis];
        }
    );
    
    size_t median = points.size() / 2;
    
    // Create node with median point
    auto node = std::make_shared<GeoKDNode>(points[median], axis);
    
    // Build left and right subtrees
    if (median > 0) {
        std::vector<std::vector<double>> left_points(points.begin(), points.begin() + median);
        node->left = build_tree(left_points, depth + 1);
    }
    
    if (median + 1 < points.size()) {
        std::vector<std::vector<double>> right_points(points.begin() + median + 1, points.end());
        node->right = build_tree(right_points, depth + 1);
    }
    
    return node;
}

void GeoKDTree::find_closest_3d(
    const std::shared_ptr<GeoKDNode>& node,
    const std::vector<double>& point,
    int depth,
    std::vector<double>& best,
    double& best_dist
) const {
    if (!node) {
        return;
    }
    
    // Calculate distance to current node (only first 3 dimensions - x, y, z)
    double node_dist = kdtree_helpers::squared_distance_3d(point, node->point);
    
    // Update best if necessary
    if (best.empty() || node_dist < best_dist) {
        best = node->point;
        best_dist = node_dist;
    }
    
    // Determine which side to search first
    int axis = node->axis;
    double diff = point[axis] - node->point[axis];
    
    auto close = (diff < 0) ? node->left : node->right;
    auto away = (diff < 0) ? node->right : node->left;
    
    // Search the close side first
    find_closest_3d(close, point, depth + 1, best, best_dist);
    
    // Check if we need to search the other side
    if (diff * diff < best_dist) {
        find_closest_3d(away, point, depth + 1, best, best_dist);
    }
}

int GeoKDTree::closest_idx(const std::pair<double, double>& point) const {
    if (!root) {
        throw std::runtime_error("GeoKDTree is empty");
    }
    
    // Convert query point to xyz
    auto query_point = kdtree_helpers::lat_lon_idx_to_xyz_idx(point.first, point.second, 0);
    
    std::vector<double> best;
    double best_dist = std::numeric_limits<double>::infinity();
    
    find_closest_3d(root, query_point, 0, best, best_dist);
    
    // Return the index (4th element)
    return static_cast<int>(best[3]);
}

ClosestIdxResult GeoKDTree::closest_idx_with_distance(const std::pair<double, double>& point) const {
    if (!root) {
        throw std::runtime_error("GeoKDTree is empty");
    }
    
    // Convert query point to xyz
    auto query_point = kdtree_helpers::lat_lon_idx_to_xyz_idx(point.first, point.second, 0);
    
    std::vector<double> best;
    double best_dist = std::numeric_limits<double>::infinity();
    
    find_closest_3d(root, query_point, 0, best, best_dist);
    
    // Return the index and distance
    return ClosestIdxResult{static_cast<int>(best[3]), best_dist};
}

std::vector<double> GeoKDTree::lat_lon_idx_to_xyz_idx(double lat, double lon, int idx) {
    return kdtree_helpers::lat_lon_idx_to_xyz_idx(lat, lon, idx);
}